Move `contiguous_strides` to `c10/util/strides.h`.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76098
* #76097
* #76096
* #77045
* __->__ #76812

This PR moves `contiguous_strides` implementation from `LinearAlgebraUtils.h` to a broader
scope: `c10/util/strides.h`. This should make it easier to reuse this implementation on
other parts of the codebase.

ysiraichi committed:
Move `contiguous_strides` to `c10/util/strides.h`.

This PR moves `contiguous_strides` implementation from `LinearAlgebraUtils.h` to a broader
scope: `c10/util/strides.h`. This should make it easier to reuse this implementation on
other parts of the codebase.

[ghstack-poisoned]

lezcano reviewed (commented):
This implementation is certainly much simpler than the one I initially wrote.

My only question is: why not work directly with `DimVector`s?

ysiraichi commented on c10/util/strides.h:
```diff
@@ -0,0 +1,27 @@
+  std::vector<Int> strides;
+
+  if (dims > 0) {
+    strides.assign(dims, 0);
```
The problem is that, on the next line, I'm populating the `dims - 1`-th element. `reserve` still requires `push_back` (or something like that) to increase the `vector` size.

Lezcano replied:
You are completely right.

ysiraichi commented on c10/util/strides.h:
```diff
@@ -0,0 +1,27 @@
+namespace c10 {
+
+// Computes the contiguous strides of a tensor, given its sizes.
+static inline std::vector<typename IntArrayRef::value_type> contiguous_strides(
```
There were a couple of reasons for that:
- `DimVector` is defined in `ATen/core/DimVector.h`. Wasn't sure I could move it to `c10`
- I could use `SmallVector` directly, but I don't think that would be good (since I would have to specify the type parameters myself)
- I didn't think it would make that much of a difference -- only a few places call `contiguous_strides`, and they only call it once or twice

Lezcano replied:
As `DimVector` is just a `using...`, probably returning a `SmallVector<typename IntArrayRef::value_type>` works just fine. This should be correct as this is the internal type that the arrayrefs of `sizes()` and `strides()` return.

That being said, yeah, it may not matter too much.

Now my question is, why is `DimVector` not in c10 @ezyang ?

ezyang replied:
It can easily go in c10, feel free to move it. Looking over the history it looks like SmallVector got moved to c10 in 2019 but we forgot to move this file too.

Returning a DimVector is better, it avoids heap allocation in the common case.

ysiraichi replied:
Implemented on #77045

ysiraichi committed:
Update on "Move `contiguous_strides` to `c10/util/strides.h`."


This PR moves `contiguous_strides` implementation from `LinearAlgebraUtils.h` to a broader
scope: `c10/util/strides.h`. This should make it easier to reuse this implementation on
other parts of the codebase.

[ghstack-poisoned]

ezyang commented:
make this non-draft?

ezyang commented:
The implementations are not obviously equivalent, do you want to run it through a model checker to show they're the same?

Lezcano commented:
It'd be better to have a look at the one in TI and see that they are equivalent, as mine is adapted from that one.

ysiraichi commented:
I modeled both implementations in `z3` (see implementation below).
Most of the code should be the same, except a few things that had to be modified in order to work with `z3`:

- Re-implement `max` for working with symbolic values
- Replacing assignment with equality. Basically, instead of assigning a value to a variable (e.g `arr[i] = 10`), we have to say that they are equal (e.g. `arr[i] == 10`). i.e. can't mutate symbolic values. 

For a given number of dimensions `dims`, the code below checks whether the new implementation results match the old implementation results for both C-contiguous and C-contiguous batches of F-contiguous matrices. I'm assuming all sizes are greater than or equal 0.

<details>
<summary>Header (utility functions)</summary>

```python
import sys
import z3

s = z3.Solver()

# Symbolic 'max' implementation
def symmax(a, b):
    return z3.If(a >= b, a, b)

# Adds a equality constraint. Replacement for assigning values.
def Eq(a, b):
    s.add(a == b)

# Adds equality for all elements of a vector
def EqVector(a, b):
    assert len(a) == len(b)
    s.add(*[a[i] == b[i] for i in range(len(a))])
```
</details>

<details>
<summary>New Implementation</summary>

```python
def new(sizes):
    strides = []
    dims = len(sizes)

    if dims > 0:
        strides = z3.IntVector("new_strides", dims)
        Eq(strides[dims - 1],
           1)
        for i in range(dims - 2, -1, -1):
            Eq(strides[i],
               strides[i + 1] * symmax(sizes[i + 1], 1))

    return strides

def new_fcontig(sizes):
    strides = new(sizes)
    dims = len(sizes)

    if dims >= 2:
        # Copying the vector, simulating "mutation". ===================
        fcontig_strides = z3.IntVector("new_fcontig_strides", dims)
        EqVector(fcontig_strides[:(dims - 2)], strides[:(dims - 2)])
        # ==============================================================
        Eq(fcontig_strides[dims - 1],
           symmax(sizes[dims - 2], 1))
        Eq(fcontig_strides[dims - 2],
           1)
        strides = fcontig_strides

    return strides
```
</details>

<details>
<summary>Old Implementation</summary>

```python

def old(sizes, f_contig):
    n = len(sizes)
    strides = z3.IntVector(f"old_strides", n)

    if n == 0:
        return []
    elif n == 1:
        Eq(strides[0],
           1)
        return strides

    last_idx = n - 1
    snd_last_idx = n - 2

    Eq(strides[snd_last_idx],
       symmax(sizes[last_idx], 1))
    for i in range(snd_last_idx - 1, -1, -1):
        Eq(strides[i],
           strides[i + 1] * symmax(sizes[i + 1], 1))

    Eq(strides[last_idx],
       symmax(sizes[snd_last_idx], 1) if f_contig else 1)

    if f_contig:
        # Copying the vector, simulating "mutation". ===================
        fcontig_strides = z3.IntVector(f"old_fcontig_strides", n)
        EqVector(fcontig_strides[:(dims - 2)], strides[:(dims - 2)])
        Eq(fcontig_strides[dims - 1], strides[dims - 1])
        # ==============================================================
        Eq(fcontig_strides[snd_last_idx],
           1)
        strides = fcontig_strides

    return strides
```
</details>

<details>
<summary>Driver Code</summary>

```python
dims = 0 if len(sys.argv) < 2 else int(sys.argv[1])
print(f">> Looking for solution for vector size: {dims}")

def solve_for(strides_fn_1, strides_fn_2):
    global s

    s = z3.Solver()
    sizes = z3.IntVector("sizes", dims)

    # Retrieve the stride vectors from the input functions.
    strides_1 = strides_fn_1(sizes)
    strides_2 = strides_fn_2(sizes)

    if dims > 0:
        # Sizes should be greater or equal than 0
        s.add(*[sizes[i] >= 0 for i in range(dims)])
        # We want to know whether any of the strides returned by the
        # two functions are different.
        s.add(z3.Or(*[strides_1[i] != strides_2[i] for i in range(dims)]))

    r = s.check()

    if r == z3.sat:
        print(">> Found one or more different strides!")
        print(s.model())
    elif r == z3.unsat:
        print(">> All strides match!")
    else:
        print(">> BAD!")

# Check whether 'contiguous_strides' implementation yields the same result.
solve_for(
    new,
    lambda sizes: old(sizes, False)
)

# Check whether 'batched_matrix_contiguous_strides' implementation yields the same result.
solve_for(
    new_fcontig,
    lambda sizes: old(sizes, True)
)
```
</details>

I've run this program until up to `dims=20`. The results for all of them showed that the results match.
Even though it serves as proof that the 2 models are the same for up to 20 dimensions, the models themselves might not exactly reflect the implementation.

ysiraichi committed:
Update on "Move `contiguous_strides` to `c10/util/strides.h`."


This PR moves `contiguous_strides` implementation from `LinearAlgebraUtils.h` to a broader
scope: `c10/util/strides.h`. This should make it easier to reuse this implementation on
other parts of the codebase.

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this

ezyang commented:
@pytorchbot revert this because it broke master build

ezyang commented:
land race I guess

ysiraichi committed:
Update on "Move `contiguous_strides` to `c10/util/strides.h`."


This PR moves `contiguous_strides` implementation from `LinearAlgebraUtils.h` to a broader
scope: `c10/util/strides.h`. This should make it easier to reuse this implementation on
other parts of the codebase.

[ghstack-poisoned]

ysiraichi committed:
Update on "Move `contiguous_strides` to `c10/util/strides.h`."


This PR moves `contiguous_strides` implementation from `LinearAlgebraUtils.h` to a broader
scope: `c10/util/strides.h`. This should make it easier to reuse this implementation on
other parts of the codebase.

[ghstack-poisoned]

ysiraichi commented:
Rebased.

ezyang commented:
@pytorchbot merge this

malfet commented:
@pytorchbot revert this as it blocks the revert of https://github.com/pytorch/pytorch/pull/72935 , which in turn breaks internal build

ysiraichi committed:
Rebased. on "Move `contiguous_strides` to `c10/util/strides.h`."


This PR moves `contiguous_strides` implementation from `LinearAlgebraUtils.h` to a broader
scope: `c10/util/strides.h`. This should make it easier to reuse this implementation on
other parts of the codebase.

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this
